File: Version4/backend/app/services/satellite_service.py
"""
Satellite data service for handling satellite imagery and data
"""

import asyncio
import aiohttp
import random
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
import os

from app.core.config import settings
from app.core.database import get_db


class SatelliteService:
    """Service for managing satellite data and imagery"""
    
    def __init__(self):
        self.db = get_db()
        self.satellite_data = {}
    
    async def get_imagery(self, satellite: str, region: str, date_range: Optional[str] = None) -> Dict[str, Any]:
        """Get satellite imagery with filters"""
        
        # Mock data for development
        images = []
        for i in range(6):
            timestamp = datetime.now() - timedelta(hours=i*2)
            images.append({
                "id": i + 1,
                "title": f"{satellite} {region.title()} View",
                "url": f"https://picsum.photos/seed/{satellite}_{region}_{i}/400/300.jpg",
                "timestamp": timestamp.isoformat(),
                "resolution": random.choice(["1km", "2km", "4km"]),
                "filename": f"{satellite}_{region}_{timestamp.strftime('%Y%m%d_%H%M')}.jpg"
            })
        
        # Mock markers for map
        markers = [
            {"lat": 40.7128, "lng": -74.0060, "title": "New York", "description": "Clear skies", "timestamp": "14:30 UTC"},
            {"lat": 34.0522, "lng": -118.2437, "title": "Los Angeles", "description": "Partly cloudy", "timestamp": "14:25 UTC"},
            {"lat": 51.5074, "lng": -0.1278, "title": "London", "description": "Overcast", "timestamp": "14:20 UTC"},
            {"lat": 35.6762, "lng": 139.6503, "title": "Tokyo", "description": "Clear", "timestamp": "14:15 UTC"},
            {"lat": -33.8688, "lng": 151.2093, "title": "Sydney", "description": "Rainy", "timestamp": "14:10 UTC"}
        ]
        
        # Mock gallery
        gallery = []
        for i in range(8):
            gallery.append({
                "id": i + 1,
                "title": f"Weather Pattern {i + 1}",
                "thumbnail": f"https://picsum.photos/seed/thumb_{i}/150/150.jpg",
                "fullSize": f"https://picsum.photos/seed/full_{i}/800/600.jpg"
            })
        
        return {
            "images": images,
            "markers": markers,
            "gallery": gallery,
            "satellite": satellite,
            "region": region,
            "total_images": len(images)
        }
    
    async def get_live_feed(self, satellite: str) -> Dict[str, str]:
        """Get live video feed URL for satellite"""
        
        # Mock video URLs - in production, these would be real satellite feeds
        video_urls = {
            "GOES-16": "https://www.w3schools.com/html/mov_bbb.mp4",
            "GOES-17": "https://www.w3schools.com/html/mov_bbb.mp4",
            "HIMAWARI-8": "https://www.w3schools.com/html/mov_bbb.mp4",
            "METEOSAT": "https://www.w3schools.com/html/mov_bbb.mp4"
        }
        
        return {
            "url": video_urls.get(satellite, video_urls["GOES-16"]),
            "satellite": satellite,
            "status": "active",
            "last_updated": datetime.now().isoformat()
        }
    
    async def get_image_by_id(self, image_id: int) -> Optional[Dict[str, Any]]:
        """Get specific image by ID"""
        
        # Mock image data
        return {
            "id": image_id,
            "url": f"https://picsum.photos/seed/image_{image_id}/800/600.jpg",
            "filename": f"satellite_image_{image_id}.jpg",
            "file_size": random.randint(500000, 2000000),  # bytes
            "timestamp": datetime.now().isoformat()
        }
    
    async def get_stats(self) -> Dict[str, Any]:
        """Get satellite processing statistics"""
        
        return {
            "active_satellites": 12,
            "total_images_processed": 45678,
            "data_processed_today": "2.4GB",
            "last_update": datetime.now().isoformat(),
            "processing_queue": 3,
            "success_rate": 98.5
        }
    
    async def health_check(self) -> str:
        """Health check for satellite service"""
        
        # In production, check actual service health
        return "healthy"
    
    async def fetch_external_data(self, satellite: str, region: str, api_source: str):
        """Fetch data from external APIs (NASA, ESA, etc.)"""
        
        # Mock external data fetching
        await asyncio.sleep(2)  # Simulate API call
        
        # In production, this would make actual API calls to NASA/ESA
        print(f"Fetching data from {api_source} for {satellite} in {region}")
        
        # Store fetched data in database
        await self.db.execute_query(
            "INSERT INTO satellite_images (satellite_name, image_url, region, resolution) VALUES (?, ?, ?, ?)",
            (satellite, f"mock_url_{datetime.now()}", region, "2km")
        )
    
    async def generate_sample_data(self):
        """Generate sample data for testing"""
        
        # Generate sample satellite images
        satellites = ["GOES-16", "GOES-17", "HIMAWARI-8", "METEOSAT"]
        regions = ["global", "north-america", "europe", "asia"]
        
        for satellite in satellites:
            for region in regions:
                for i in range(5):
                    await self.db.execute_query(
                        """INSERT INTO satellite_images 
                           (satellite_name, image_url, timestamp, region, resolution, file_size) 
                           VALUES (?, ?, ?, ?, ?, ?)""",
                        (
                            satellite,
                            f"https://picsum.photos/seed/{satellite}_{region}_{i}/400/300.jpg",
                            datetime.now() - timedelta(hours=i*6),
                            region,
                            random.choice(["1km", "2km", "4km"]),
                            random.randint(500000, 2000000)
                        )
                    )
        
        print("✅ Sample satellite data generated")

File: Version4/frontend/src/services/satelliteService.js
import axios from 'axios';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

const satelliteService = {
  // Get dashboard statistics
  getStats: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/stats`);
      return response.data;
    } catch (error) {
      console.error('Error fetching stats:', error);
      // Return mock data for development
      return {
        activeSatellites: 12,
        predictionsToday: 248,
        dataProcessed: 1.2,
        modelAccuracy: 94.5
      };
    }
  },

  // Get recent predictions
  getRecentPredictions: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/predictions/recent`);
      return response.data;
    } catch (error) {
      console.error('Error fetching recent predictions:', error);
      // Return mock data
      return [
        {
          id: 1,
          location: 'North America',
          timestamp: '2024-01-20 14:30:00',
          accuracy: 94,
          status: 'completed'
        },
        {
          id: 2,
          location: 'Europe',
          timestamp: '2024-01-20 14:15:00',
          accuracy: 91,
          status: 'completed'
        },
        {
          id: 3,
          location: 'Asia Pacific',
          timestamp: '2024-01-20 14:00:00',
          accuracy: 89,
          status: 'processing'
        }
      ];
    }
  },

  // Get chart data for dashboard
  getChartData: async () => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/charts/performance`);
      return response.data;
    } catch (error) {
      console.error('Error fetching chart data:', error);
      // Return mock chart data
      return {
        labels: ['00:00', '04:00', '08:00', '12:00', '16:00', '20:00'],
        datasets: [
          {
            label: 'Generator Loss',
            data: [0.5, 0.45, 0.4, 0.35, 0.3, 0.25],
            borderColor: 'rgb(75, 192, 192)',
            backgroundColor: 'rgba(75, 192, 192, 0.2)',
            tension: 0.1
          },
          {
            label: 'Discriminator Loss',
            data: [0.6, 0.55, 0.5, 0.4, 0.35, 0.3],
            borderColor: 'rgb(255, 99, 132)',
            backgroundColor: 'rgba(255, 99, 132, 0.2)',
            tension: 0.1
          }
        ]
      };
    }
  },

  // Get satellite imagery
  getSatelliteImagery: async (params) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/satellite/imagery`, { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching satellite imagery:', error);
      // Return mock data
      return {
        images: [
          {
            id: 1,
            title: 'GOES-16 Full Disk',
            url: 'https://picsum.photos/seed/sat1/400/300.jpg',
            timestamp: '2024-01-20 14:30:00',
            resolution: '2km',
            filename: 'goes16_full_disk_20240120_1430.jpg'
          },
          {
            id: 2,
            title: 'North America View',
            url: 'https://picsum.photos/seed/sat2/400/300.jpg',
            timestamp: '2024-01-20 14:25:00',
            resolution: '1km',
            filename: 'north_america_20240120_1425.jpg'
          },
          {
            id: 3,
            title: 'Pacific Region',
            url: 'https://picsum.photos/seed/sat3/400/300.jpg',
            timestamp: '2024-01-20 14:20:00',
            resolution: '4km',
            filename: 'pacific_region_20240120_1420.jpg'
          }
        ],
        markers: [
          { lat: 40.7128, lng: -74.0060, title: 'New York', description: 'Clear skies', timestamp: '14:30 UTC' },
          { lat: 34.0522, lng: -118.2437, title: 'Los Angeles', description: 'Partly cloudy', timestamp: '14:25 UTC' },
          { lat: 51.5074, lng: -0.1278, title: 'London', description: 'Overcast', timestamp: '14:20 UTC' }
        ],
        gallery: [
          {
            id: 1,
            title: 'Hurricane Formation',
            thumbnail: 'https://picsum.photos/seed/thumb1/150/150.jpg',
            fullSize: 'https://picsum.photos/seed/full1/800/600.jpg'
          },
          {
            id: 2,
            title: 'Cloud Patterns',
            thumbnail: 'https://picsum.photos/seed/thumb2/150/150.jpg',
            fullSize: 'https://picsum.photos/seed/full2/800/600.jpg'
          }
        ]
      };
    }
  },

  // Get live video feed
  getLiveVideoFeed: async (satellite) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/satellite/live-feed`, {
        params: { satellite }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching live video feed:', error);
      // Return mock video URL
      return {
        url: 'https://www.w3schools.com/html/mov_bbb.mp4'
      };
    }
  },

  // Download satellite image
  downloadImage: async (imageId, filename) => {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/satellite/download/${imageId}`, {
        responseType: 'blob'
      });
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', filename);
      document.body.appendChild(link);
      link.click();
      link.remove();
    } catch (error) {
      console.error('Error downloading image:', error);
    }
  }
};

export default satelliteService;
